src/Integration/QualityProfiles/SonarQubeQualityProfileProvider.cs

```diff
@@ -0,0 +1,152 @@
+        }
+
+        private async Task<RoslynExportProfileResponse> GetSonarQubeQualityProfile(BoundSonarQubeProject project, Language language)
+        {
+            var serverLanguage = language.ToServerLanguage();
+
+            var qualityProfileInfo = await SafeServiceCall(() => this.sonarQubeService.GetQualityProfileAsync(
+                project.ProjectKey, project.Organization?.Key, serverLanguage, CancellationToken.None));
+            if (qualityProfileInfo == null)
+            {
+                this.logger.WriteLine(string.Format(Strings.SubTextPaddingFormat,
+                   string.Format(Strings.CannotDownloadQualityProfileForLanguage, language.Name)));
+                return null;
+            }
+
+            var roslynProfileExporter = await SafeServiceCall(() => this.sonarQubeService.GetRoslynExportProfileAsync(
+                qualityProfileInfo.Name, project.Organization?.Key, serverLanguage, CancellationToken.None));
+            if (roslynProfileExporter == null)
+            {
+                this.logger.WriteLine(string.Format(Strings.SubTextPaddingFormat,
+                    string.Format(Strings.QualityProfileDownloadFailedMessageFormat, qualityProfileInfo.Name,
+                        qualityProfileInfo.Key, language.Name)));
+                return null;
+            }
+
+            return roslynProfileExporter;
+        }
+
+        private async Task<T> SafeServiceCall<T>(Func<Task<T>> call)
+        {
+            try
+            {
+                return await call();
+            }
+            catch (HttpRequestException e)
+            {
+                // For some errors we will get an inner exception which will have a more specific information
+                // that we would like to show i.e.when the host could not be resolved
+                var innerException = e.InnerException as System.Net.WebException;
+                this.logger.WriteLine(Strings.SonarQubeRequestFailed, e.Message, innerException?.Message);
+            }
+            catch (TaskCanceledException)
+            {
+                // Canceled or timeout
+                this.logger.WriteLine(Strings.SonarQubeRequestTimeoutOrCancelled);
+            }
+            catch (Exception ex)
+            {
+                this.logger.WriteLine(Strings.SonarQubeRequestFailed, ex.Message, null);
+            }
+
+            return default(T);
+        }
+
+        private RuleSet ConvertToCodeAnalysisRuleSet(RoslynExportProfileResponse roslynProfileExporter)
+        {
+            var tempRuleSetFilePath = Path.GetTempFileName();
+            File.WriteAllText(tempRuleSetFilePath, roslynProfileExporter.Configuration.RuleSet.OuterXml);
+
+            return RuleSet.LoadFromFile(tempRuleSetFilePath);
+        }
+
+        private QualityProfile ConvertToSonarLintQualityProfile(RuleSet ruleset, Language language)
+        {
+            var rules = ruleset.Rules.Select(r => new SonarRule(r.RuleInfo.AnalyzerId, true));
```

Comment:
Should the last parameter always be _true_ - doesn't the ruleset also contain info about rules are that not enabled?

Reply:
No it shouldn't I put that to have something compiling and forgot to modify it. I need to find how to detect whether a rule is enabled or not.